x/incentives/keeper/gauge_asset.go

```diff
@@ -0,0 +1,191 @@
+		d.idToDecodedAddr = append(d.idToDecodedAddr, decodedOwnerAddr)
+		d.idToDistrCoins = append(d.idToDistrCoins, rewards)
+	}
+	return nil
+}
+
+// sendRewardsToLocks utilizes provided distributionInfo to send coins from the module account to various recipients.
+func (k Keeper) sendRewardsToLocks(ctx sdk.Context, distrs *distributionInfo) error {
+	numIDs := len(distrs.idToDecodedAddr)
+	if len(distrs.idToDistrCoins) != numIDs {
+		return fmt.Errorf("number of addresses and coins to distribute to must be equal")
+	}
+	ctx.Logger().Debug(fmt.Sprintf("Beginning distribution to %d users", numIDs))
+
+	for id := 0; id < numIDs; id++ {
+		err := k.bk.SendCoinsFromModuleToAccount(
+			ctx,
+			types.ModuleName,
+			distrs.idToDecodedAddr[id],
+			distrs.idToDistrCoins[id])
+		if err != nil {
+			return err
+		}
+	}
+	ctx.Logger().Debug("Finished sending, now creating liquidity add events")
+	for id := 0; id < numIDs; id++ {
+		ctx.EventManager().EmitEvents(sdk.Events{
+			sdk.NewEvent(
+				types.TypeEvtDistribution,
+				sdk.NewAttribute(types.AttributeReceiver, distrs.idToBech32Addr[id]),
+				sdk.NewAttribute(types.AttributeAmount, distrs.idToDistrCoins[id].String()),
+			),
+		})
+	}
+	ctx.Logger().Debug(fmt.Sprintf("Finished Distributing to %d users", numIDs))
+	return nil
+}
+
+// distributeToAssetGauge runs the distribution logic for a gauge, and adds the sends to
+// the distrInfo struct. It also updates the gauge for the distribution.
+// Locks is expected to be the correct set of lock recipients for this gauge.
+func (k Keeper) distributeToAssetGauge(ctx sdk.Context, gauge types.Gauge, currResult *distributionInfo) (sdk.Coins, error) {
+	assetDist := gauge.GetAsset()
+	if assetDist == nil {
+		return sdk.Coins{}, fmt.Errorf("gauge %d is not an asset gauge", gauge.Id)
+	}
+
+	locksByDenomCache := make(map[string][]lockuptypes.PeriodLock)
+	locks := k.getDistributeToBaseLocks(ctx, gauge, locksByDenomCache)
+
+	denom := assetDist.Denom
+	lockSum := lockuptypes.SumLocksByDenom(locks, denom)
+
+	if lockSum.IsZero() {
+		return sdk.Coins{}, nil
+	}
+
+	remainCoins := gauge.Coins.Sub(gauge.DistributedCoins...)
+	// if its a perpetual gauge, we set remaining epochs to 1.
+	// otherwise is is a non perpetual gauge and we determine how many epoch payouts are left
+	remainEpochs := uint64(1)
+	if !gauge.IsPerpetual {
+		remainEpochs = gauge.NumEpochsPaidOver - gauge.FilledEpochs
+	}
+
+	/* ---------------------------- defense in depth ---------------------------- */
+	// this should never happen in practice since gauge passed in should always be an active gauge.
+	if remainEpochs == uint64(0) {
+		ctx.Logger().Error(fmt.Sprintf("gauge %d has no remaining epochs, skipping", gauge.Id))
+		return sdk.Coins{}, nil
+	}
+
+	// this should never happen in practice
+	if remainCoins.Empty() {
+		ctx.Logger().Error(fmt.Sprintf("gauge %d is empty, skipping", gauge.Id))
+		err := k.updateGaugePostDistribute(ctx, gauge, sdk.Coins{})
+		return sdk.Coins{}, err
```

Comment:
dont log and return error

Reply:
the `err` is for the `updateGaugePostDistribute`